internal/builder/builder.go

```diff
@@ -0,0 +1,279 @@
+
+	apiv1 "github.com/substratusai/substratus/api/v1"
+	"github.com/substratusai/substratus/internal/cloud"
+	"github.com/substratusai/substratus/internal/resources"
+	batchv1 "k8s.io/api/batch/v1"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	ptr "k8s.io/utils/pointer"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+const (
+	saName = "container-builder"
+)
+
+type BuildableObject interface {
+	client.Object
+	GetContainer() apiv1.Container
+	GetConditions() *[]metav1.Condition
+}
+
+type Result struct {
+	ctrl.Result
+	Complete bool
+}
+
+type ContainerReconciler struct {
+	Scheme *runtime.Scheme
+	Client client.Client
+
+	CloudContext *cloud.Context
+
+	Kind string
+}
+
+func (r *ContainerReconciler) ReconcileContainer(ctx context.Context, obj BuildableObject) (Result, error) {
+	log := log.FromContext(ctx)
+
+	if obj.GetContainer().Image != "" {
+		return Result{Complete: true}, nil
+	}
+
+	log.Info("Reconciling container")
+	defer log.Info("Done reconciling container")
+
+	// Service account used for building and pushing the image.
+	sa := &corev1.ServiceAccount{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      saName,
+			Namespace: obj.GetNamespace(),
+		},
+	}
+	if _, err := controllerutil.CreateOrUpdate(ctx, r.Client, sa, func() error {
+		return r.authNServiceAccount(sa)
+	}); err != nil {
+		return Result{}, fmt.Errorf("failed to create or update service account: %w", err)
+	}
+
+	// The Job that will build the container image.
+	buildJob, err := r.buildJob(ctx, obj)
+	if err != nil {
+		log.Error(err, "unable to construct image-builder Job")
+		// No use in retrying...
+		return Result{}, nil
+	}
+
+	if err := r.Client.Get(ctx, client.ObjectKeyFromObject(buildJob), buildJob); err != nil {
+		if apierrors.IsNotFound(err) {
+			if err := r.Client.Create(ctx, buildJob); client.IgnoreAlreadyExists(err) != nil {
+				return Result{}, fmt.Errorf("creating builder Job: %w", err)
+			}
+		} else {
+			return Result{}, fmt.Errorf("getting builder Job: %w", err)
+		}
+	}
+
+	if buildJob.Status.Succeeded < 1 {
+		log.Info("The builder Job has not succeeded yet")
+
+		meta.SetStatusCondition(obj.GetConditions(), metav1.Condition{
+			Type:               apiv1.ConditionBuilt,
+			Status:             metav1.ConditionFalse,
+			Reason:             apiv1.ReasonJobNotComplete,
+			ObservedGeneration: obj.GetGeneration(),
+			Message:            fmt.Sprintf("Waiting for builder Job to complete: %v", buildJob.Name),
+		})
+		if err := r.Client.Status().Update(ctx, obj); err != nil {
+			return Result{}, fmt.Errorf("updating status: %w", err)
+		}
+
+		// Allow Job watch to requeue.
+		return Result{}, nil
+	}
+
+	container := obj.GetContainer()
```

Comment:
GetContainer returns a copy of the Container object so the next few lines shouldn't be able to update the actual Image attribute of the underlying `obj`?